dhis-2/dhis-support/dhis-support-hibernate/src/main/java/org/hisp/dhis/config/DataSourceConfig.java

```diff
@@ -65,28 +65,76 @@ public class DataSourceConfig {
+  public NamedParameterJdbcTemplate namedParameterJdbcTemplate(
+      @Qualifier("actualDataSource") DataSource dataSource) {
     return new NamedParameterJdbcTemplate(dataSource);
   }
 
   @Primary
   @Bean
-  public JdbcTemplate jdbcTemplate(DataSource dataSource) {
+  public JdbcTemplate jdbcTemplate(@Qualifier("actualDataSource") DataSource dataSource) {
     JdbcTemplate jdbcTemplate = new JdbcTemplate(dataSource);
     jdbcTemplate.setFetchSize(1000);
     return jdbcTemplate;
   }
 
-  @Bean
+  @Bean("readOnlyNamedParameterJdbcTemplate")
+  public NamedParameterJdbcTemplate readOnlyNamedParameterJdbcTemplate(
+      @Qualifier("readReplicaDataSource") DataSource dataSource) {
+    return new NamedParameterJdbcTemplate(dataSource);
+  }
+
+  @Bean("readOnlyJdbcTemplate")
   public JdbcTemplate readOnlyJdbcTemplate(
-      DhisConfigurationProvider config, DataSource dataSource) {
-    ReadOnlyDataSourceManager manager = new ReadOnlyDataSourceManager(config);
+      @Qualifier("readReplicaDataSource") DataSource dataSource) {
+    JdbcTemplate readOnlyJdbcTemplate = new JdbcTemplate(dataSource);
+    readOnlyJdbcTemplate.setFetchSize(1000);
 
-    JdbcTemplate jdbcTemplate =
-        new JdbcTemplate(MoreObjects.firstNonNull(manager.getReadOnlyDataSource(), dataSource));
-    jdbcTemplate.setFetchSize(1000);
+    return readOnlyJdbcTemplate;
+  }
 
-    return jdbcTemplate;
+  @Bean("readReplicaDataSource")
+  public DataSource readReplicaDataSource(
+      DhisConfigurationProvider config,
+      @Qualifier("actualDataSource") DataSource actualDataSource) {
+
+    if (!config.isTrackerReadReplicaEnabled()) {
+      log.info("Tracker Read replica is disabled");
+      return actualDataSource;
+    }
+    String jdbcUrl = config.getProperty(ConfigurationKey.TRACKER_READ_REPLICA_CONNECTION_URL);
+    String username = config.getProperty(ConfigurationKey.CONNECTION_USERNAME);
+
+    if (!StringUtils.hasText(jdbcUrl)) {
+      String errorMessage =
+          "Tracker Read replica is enabled but read replica connection url not specified in dhis.conf.";
+      log.error(errorMessage);
+      throw new IllegalStateException(errorMessage);
+    }
+    log.info("Tracker Read replica is enabled");
+    String dbPoolType = config.getProperty(ConfigurationKey.DB_POOL_TYPE);
+
+    DbPoolConfig poolConfig =
+        DbPoolConfig.builder()
+            .dhisConfig(config)
+            .jdbcUrl(jdbcUrl)
+            .username(username)
+            .dbPoolType(dbPoolType)
+            .build();
+
+    try {
+      return DatabasePoolUtils.createDbPool(poolConfig);
+    } catch (SQLException | PropertyVetoException e) {
+      String message =
+          String.format(
+              "Connection test failed for tracker read replica database pool, jdbcUrl: '%s', user: '%s'",
+              jdbcUrl, username);
+
+      log.error(message);
+      log.error(DebugUtils.getStackTrace(e));
```

Comment:
Some general things on logging

1. We should handle an error once. That means either log and move on or throw.
2. The logger can log exceptions so do not log twice but use the appropriate method.
3. Logs should be actionable. Think of the admin that will then have to deal with multiple log entries that are all pointing to the same problem. Is that helpful?

For this particular case: I assume you want to prevent DHIS2 from starting and give a meaningful message so an admin can fix the config right? Then rethrow with a refined message.

Right now this leads to at least 2 log entries for your two `log.error` calls + likely another one higher up the stack for the exception you rethrow.